### PR TITLE
fix: remove extra path in console local dir

### PR DESCRIPTION
**Pull request by Xunzhuo**
Signed-off-by: Xunzhuo <mixdeers@gmail.com>

**Describe the change**

Minor fix to the console path

Follow these commands in README:

``` bash
# Checkout the source code
mkdir kiali_sources
cd kiali_sources
export KIALI_SOURCES=$(realpath .)

git clone https://github.com/kiali/kiali.git
git clone https://github.com/kiali/kiali-ui.git
git clone https://github.com/kiali/kiali-operator.git
git clone https://github.com/kiali/helm-charts.git

ln -s $KIALI_SOURCES/kiali-operator kiali/operator

# Build the back-end and run the tests
cd $KIALI_SOURCES/kiali
make build test

# Build the front-end and run the tests
cd $KIALI_SOURCES/kiali-ui
yarn && yarn build && yarn test
```

When we run `make container-build` :

```
❯ make container-build
CONSOLE_DIR is /root/github.com/kiali_sources/kiali/../../../../../kiali-ui
Apparently, this CONSOLE_DIR does not contain the kiali-ui
make: *** [make/Makefile.container.mk:7: .prepare-kiali-image-files] Error 1
```
The default path should be:

```
CONSOLE_LOCAL_DIR ?= ${ROOTDIR}/../kiali-ui
```

### CODE REVIEW

**Comment by jmazzitelli** [comment]
No, that's not how some (many) have this checked out... at least legacy devs from older go environments.

We have this checked out at `$HOME/source/kiali/kiali/src/github.com/kiali/kiali` which is why it is like it is.

Let me see if its easy enough for those legacy devs to convert, before we switch this.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
The README looks wrong. Because when I do it, I get a build error. I suspect we need to tell people what their GOPATH should be, but I'm not sure.

So, unless someone tells me how to get my dev environment to work by following the README, I'll assume the Makefile is fine, but the README needs to be corrected :) 

```
$ pwd
/home/jmazzite/source/kiali/kiali

$ make clean
$GOPATH/go.mod exists but should not
$GOPATH/go.mod exists but should not
Cleaning...

$ make build
$GOPATH/go.mod exists but should not
$GOPATH/go.mod exists but should not
$GOPATH/go.mod exists but should not
$GOPATH/go.mod exists but should not
Building...
GOOS= GOARCH= CGO_ENABLED=0  go build \
	-o /home/jmazzite/source/kiali/kiali/bin/kiali -ldflags "-X main.version=v1.47.0-SNAPSHOT -X main.commitHash=116b20ee1d3ada8d4040dd7ca8ed018fe62e8064"
$GOPATH/go.mod exists but should not
make: *** [make/Makefile.build.mk:25: build] Error 1
```

**Comment by Xunzhuo** [comment]
@jmazzitelli I will reach it out to fix it.

**Comment by Xunzhuo** [comment]
Follow developer setup in README https://github.com/kiali/kiali#developer-setup: 
Run these commands:

```bash
# Checkout the source code
mkdir kiali_sources
cd kiali_sources
export KIALI_SOURCES=$(realpath .)

git clone https://github.com/kiali/kiali.git
git clone https://github.com/kiali/kiali-ui.git
git clone https://github.com/kiali/kiali-operator.git
git clone https://github.com/kiali/helm-charts.git

ln -s $KIALI_SOURCES/kiali-operator kiali/operator

# Build the back-end and run the tests
cd $KIALI_SOURCES/kiali
make build test

# Build the front-end and run the tests
cd $KIALI_SOURCES/kiali-ui
yarn && yarn build && yarn test
```

I think everything looks fine.

``` bash
❯ pwd
/root/kiali_sources
❯ ls
helm-charts  kiali  kiali-operator  kiali-ui
❯ cd kiali && make build test
Building...
GOOS=linux GOARCH=amd64 CGO_ENABLED=0  go build \
        -o /root/go/bin/kiali -ldflags "-X main.version=v1.47.0-SNAPSHOT -X main.commitHash=116b20ee1d3ada8d4040dd7ca8ed018fe62e8064"
Running tests, excluding third party tests under vendor
go test github.com/kiali/kiali github.com/kiali/kiali/appstate github.com/kiali/kiali/business github.com/kiali/kiali/business/authentication github.com/kiali/kiali/business/checkers github.com/kiali/kiali/business/checkers/authorization github.com/kiali/kiali/business/checkers/common github.com/kiali/kiali/business/checkers/destinationrules github.com/kiali/kiali/business/checkers/gateways github.com/kiali/kiali/business/checkers/peerauthentications github.com/kiali/kiali/business/checkers/serviceentries github.com/kiali/kiali/business/checkers/services github.com/kiali/kiali/business/checkers/sidecars github.com/kiali/kiali/business/checkers/virtualservices github.com/kiali/kiali/business/checkers/workloads github.com/kiali/kiali/config github.com/kiali/kiali/config/dashboards github.com/kiali/kiali/config/security github.com/kiali/kiali/graph github.com/kiali/kiali/graph/api github.com/kiali/kiali/graph/config/cytoscape github.com/kiali/kiali/graph/telemetry github.com/kiali/kiali/graph/telemetry/istio github.com/kiali/kiali/graph/telemetry/istio/appender github.com/kiali/kiali/graph/telemetry/istio/util github.com/kiali/kiali/handlers github.com/kiali/kiali/jaeger github.com/kiali/kiali/jaeger/jaegertest github.com/kiali/kiali/jaeger/model github.com/kiali/kiali/jaeger/model/converter/json github.com/kiali/kiali/jaeger/model/json github.com/kiali/kiali/kubernetes github.com/kiali/kiali/kubernetes/cache github.com/kiali/kiali/kubernetes/kubetest github.com/kiali/kiali/log github.com/kiali/kiali/models github.com/kiali/kiali/observability github.com/kiali/kiali/prometheus github.com/kiali/kiali/prometheus/internalmetrics github.com/kiali/kiali/prometheus/prometheustest github.com/kiali/kiali/routing github.com/kiali/kiali/server github.com/kiali/kiali/status github.com/kiali/kiali/tests/data github.com/kiali/kiali/tests/testutils/validations github.com/kiali/kiali/util github.com/kiali/kiali/util/grpcutil github.com/kiali/kiali/util/httputil github.com/kiali/kiali/util/intutil github.com/kiali/kiali/util/mtls
ok      github.com/kiali/kiali  (cached)
?       github.com/kiali/kiali/appstate [no test files]
ok      github.com/kiali/kiali/business (cached)
ok      github.com/kiali/kiali/business/authentication  (cached)
ok      github.com/kiali/kiali/business/checkers        (cached)
ok      github.com/kiali/kiali/business/checkers/authorization  (cached)
ok      github.com/kiali/kiali/business/checkers/common (cached)
ok      github.com/kiali/kiali/business/checkers/destinationrules       (cached)
ok      github.com/kiali/kiali/business/checkers/gateways       (cached)
ok      github.com/kiali/kiali/business/checkers/peerauthentications    (cached)
ok      github.com/kiali/kiali/business/checkers/serviceentries (cached)
ok      github.com/kiali/kiali/business/checkers/services       (cached)
ok      github.com/kiali/kiali/business/checkers/sidecars       (cached)
ok      github.com/kiali/kiali/business/checkers/virtualservices        (cached)
ok      github.com/kiali/kiali/business/checkers/workloads      (cached)
ok      github.com/kiali/kiali/config   (cached)
ok      github.com/kiali/kiali/config/dashboards        (cached)
ok      github.com/kiali/kiali/config/security  (cached)
?       github.com/kiali/kiali/graph    [no test files]
ok      github.com/kiali/kiali/graph/api        (cached)
ok      github.com/kiali/kiali/graph/config/cytoscape   (cached)
?       github.com/kiali/kiali/graph/telemetry  [no test files]
?       github.com/kiali/kiali/graph/telemetry/istio    [no test files]
ok      github.com/kiali/kiali/graph/telemetry/istio/appender   (cached)
?       github.com/kiali/kiali/graph/telemetry/istio/util       [no test files]
ok      github.com/kiali/kiali/handlers (cached)
?       github.com/kiali/kiali/jaeger   [no test files]
?       github.com/kiali/kiali/jaeger/jaegertest        [no test files]
?       github.com/kiali/kiali/jaeger/model     [no test files]
?       github.com/kiali/kiali/jaeger/model/converter/json      [no test files]
?       github.com/kiali/kiali/jaeger/model/json        [no test files]
ok      github.com/kiali/kiali/kubernetes       (cached)
ok      github.com/kiali/kiali/kubernetes/cache (cached)
?       github.com/kiali/kiali/kubernetes/kubetest      [no test files]
ok      github.com/kiali/kiali/log      (cached)
ok      github.com/kiali/kiali/models   (cached)
ok      github.com/kiali/kiali/observability    (cached)
?       github.com/kiali/kiali/prometheus       [no test files]
?       github.com/kiali/kiali/prometheus/internalmetrics       [no test files]
ok      github.com/kiali/kiali/prometheus/prometheustest        (cached)
ok      github.com/kiali/kiali/routing  0.053s
ok      github.com/kiali/kiali/server   1.488s
ok      github.com/kiali/kiali/status   (cached)
?       github.com/kiali/kiali/tests/data       [no test files]
ok      github.com/kiali/kiali/tests/testutils/validations      (cached)
ok      github.com/kiali/kiali/util     (cached)
?       github.com/kiali/kiali/util/grpcutil    [no test files]
ok      github.com/kiali/kiali/util/httputil    (cached)
?       github.com/kiali/kiali/util/intutil     [no test files]
?       github.com/kiali/kiali/util/mtls        [no test files]
```

**Comment by jmazzitelli** [comment]
What is your version of Go and what is your GOPATH set to?

**Comment by Xunzhuo** [comment]
```
❯ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/root/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/root/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.17.6"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1185502165=/tmp/go-build -gno-record-gcc-switches"
```

**Comment by Xunzhuo** [comment]
let us keep this PR or close this, I think this is just a minor fix about dev setup, we do not need to spend more time on it :)

**Comment by jmazzitelli** [comment]
I'm going to look into this. I have wanted to get away from the setup that requires the kiali repo git clone to go inside `src/github.com/kiali` (which is what this Makefile assumes and how many devs have it set up, including me). This was done because, back in the day, Go and GOPATH behaved differently. I also need to look at the rest of the hack scripts and Makefiles to ensure nothing breaks with this change.

So, I'm leaving this PR open; I'll be looking into all of this.

**Comment by Xunzhuo** [comment]
True, cloning into src/github.com/kiali is inconvenient. Rely on Go Module, I think we can clone Kiali repos to anywhere we want. It will help a lot to developers environment setup.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
So long as a person has a correct GOPATH, things should be working. However, @Xunzhuo you need to update this as part of this PR:

```diff
diff --git a/deploy/get-console.sh b/deploy/get-console.sh
index c2013ede..af6a631b 100755
--- a/deploy/get-console.sh
+++ b/deploy/get-console.sh
@@ -8,7 +8,7 @@
 # See the main Makefile for more info.
 
 DIR=$(dirname $0)/..
-CONSOLE_DIR=${CONSOLE_LOCAL_DIR:-$DIR/../../../../../kiali-ui}
+CONSOLE_DIR=${CONSOLE_LOCAL_DIR:-$DIR/../kiali-ui}
```

and 

```diff
diff --git a/hack/ci-openshift-molecule-tests.sh b/hack/ci-openshift-molecule-tests.sh
index d1d9bffe..ff4a56df 100755
--- a/hack/ci-openshift-molecule-tests.sh
+++ b/hack/ci-openshift-molecule-tests.sh
@@ -390,7 +390,7 @@ infomsg "Log into the cluster [${OPENSHIFT_API}] as kubeadmin user named [${KUBE
 $OC login -u ${KUBEADMIN_USER} -p ${KUBEADMIN_PW} ${OPENSHIFT_API}
 
 if [ "${USE_DEV_IMAGES}" == "true" ]; then
-  GOPATH="${SRC}/kiali/src/github.com/kiali/kiali"
+  GOPATH="${GOPATH:-/tmp}"
   infomsg "Dev images are to be tested. Will prepare them now using GOPATH=${GOPATH}"
```

@israel-hdez can you tell me if anything has to change in the Jenkinsfile? I don't think so, but I ask because I see `src/github.com/kiali` as the place where both `kiali` and `kiali-ui` are git cloned (e.g. https://github.com/kiali/kiali/blob/master/deploy/jenkins-ci/Jenkinsfile#L124). Any place where I see `src/github.com/kiali` in our scripts I want to examine closely to see if this change to the default CONSOLE_LOCAL_DIR will break. Again, I don't think this does break the Jenkinsfile, but I just want another set of eyes.

I examined the other places where this might cause breakages (the hack/ci-* files for example) but it seems like we explicitly define CONSOLE_LOCAL_DIR and thus do not assume the default location - so those all look OK. I do not see any other places that might break other than the Jenkinsfile (I just need @israel-hdez to confirm what I think - that this is OK) and diffs I post above.

**Comment by israel-hdez** [comment]
> @israel-hdez can you tell me if anything has to change in the Jenkinsfile? I don't think so

I also don't think anything needs to be changed there. There is a helper Makefile for Jenkins that already sets it's own CONSOLE_LOCAL_DIR here: https://github.com/kiali/kiali/blob/6ade22a7ae0df5cb51db53a3b19e0d2bbea467a0/deploy/jenkins-ci/Makefile#L5

File: README.adoc
Comment (jmazzitelli) [reply README.adoc:65]: based on his commit comment, looks like he is on MacOS which doesn't have realpath by default (you need to install a separate package for it). See: https://stackoverflow.com/questions/3572030/bash-script-absolute-path-with-os-x

Comment (israel-hdez) [reply README.adoc:65]: OK, it's something to keep an eye on if something breaks, because both commands have different behavior when dealing with symlinks.

Comment (jmazzitelli) [reply README.adoc:65]: this is only the README - so not affecting anything in our hack scripts or makefiles.

Comment (israel-hdez) [reply README.adoc:65]: > this is only the README - so not affecting anything in our hack scripts or makefiles.

Well, that's true :-)

**Review by jmazzitelli — APPROVED** [review]

File: README.adoc
Comment (Xunzhuo) [reply README.adoc:65]: True, `pwd` is both supported in MACOS and Linux distribution, just for compatibility purpose to make this change.

Comment (Xunzhuo) [reply README.adoc:65]: For those who do not know how to install realpath in macOS, `pwd` is a better choice, we do not need macOS users to intall an extra command.

**Comment by israel-hdez** [comment]
@jmazzitelli I wonder if we can merge this, because the `do not merge` label is still set.

**Comment by jmazzitelli** [comment]
sorry - yes, we can merge.

**Pull request merged**
